packages/aws-cdk/lib/contextplugins.ts

```diff
@@ -48,21 +53,85 @@ export class SSMContextProviderPlugin implements ContextProviderPlugin {
   }
 }
 
+export class HostedZoneContextProviderPlugin implements ContextProviderPlugin {
+
+  constructor(private readonly aws: SDK) {
+  }
+
+  public async getValue(filter: cxapi.ContextProviderProps) {
+    const account = filter.account;
+    const region = filter.region;
+    if (!this.isHostedZoneProps(filter.props)) {
+      throw new Error(`HostedZoneProvider requires domainName property to be set in ${filter.props}`);
+    }
+    const props: cxapi.HostedZoneProviderProps = filter.props;
+    const domainName = props.domainName;
+    const privateZone: boolean = !!props.privateZone;
+    const vpcId =
+    debug(`Reading hosted zone ${account}:${region}:${domainName}`);
+    const r53 = await this.aws.route53(account, region, Mode.ForReading);
+    const response = await r53.listHostedZonesByName({ DNSName: domainName }).promise();
+    if (!response.HostedZones) {
+      throw new Error(`Hosted Zone not found in account ${account}, region ${region}: ${domainName}`);
+    }
+    const candidateZones = this.filterZones(r53, response.HostedZones,
+      !!props.privateZone, props.vpcId);
+    if (candidateZones.length > 1) {
+      const filteProps = `dns:${domainName}, privateZone:${privateZone}, vpcId:${vpcId}`;
+      throw new Error(`Found more than one matching HostedZone ${candidateZones} for ${filteProps}`);
+    }
+    return candidateZones[0];
+  }
+
+  private filterZones(r53: AWS.Route53, zones: AWS.Route53.HostedZone[], privateZone: boolean, vpcId: string | undefined): AWS.Route53.HostedZone[] {
+
+      let candidates: AWS.Route53.HostedZone[] = [];
+      if (privateZone) {
+        candidates = zones.filter(zone => zone.Config && zone.Config.PrivateZone);
+      } else {
+        candidates = zones.filter(zone => !zone.Config || !zone.Config.PrivateZone);
+      }
+      if (vpcId) {
+        const vpcZones: AWS.Route53.HostedZone[] = [];
+        for (const zone of candidates) {
+          r53.getHostedZone({Id: zone.Id}, (err, data) => {
```

Comment:
This is not going to do what you want, because it's going to be asynchronous.

Just add an `await`.